packages/salesforce-adapter/src/additional_references.ts

```diff
@@ -86,147 +103,225 @@ const createReferenceMapping = (
   // ref: https://ideas.salesforce.com/s/idea/a0B8W00000GdlSPUAZ/permission-sets-with-page-layout-assignment
-  const layoutRefs = awu(addedOrModifiedInstancesChangesByType[LAYOUT_TYPE_ID_METADATA_TYPE] ?? [])
-    .flatMap(async layout => refsFromProfileOrPermissionSet(profilesAndPermSetsChanges, layout, LAYOUTS_SECTION))
+  const layoutRefs = await instanceRefsFromProfileOrPermissionSet(
+    profilesAndPermSetsChanges,
+    LAYOUTS_SECTION,
+    layoutAssignment => ({ typeName: LAYOUT_TYPE_ID_METADATA_TYPE, refName: layoutAssignment[0].layout }),
+    instancesIndex.byTypeAndApiName,
+  )
 
-  const apexPageRefs = awu(addedOrModifiedInstancesChangesByType[APEX_PAGE_METADATA_TYPE] ?? [])
-    .flatMap(async apexPage => refsFromProfileOrPermissionSet(profilesAndPermSetsChanges, apexPage, APEX_PAGE_SECTION))
+  const apexPageRefs = await instanceRefsFromProfileOrPermissionSet(
+    profilesAndPermSetsChanges,
+    APEX_PAGE_SECTION,
+    apexPage => ({ typeName: APEX_PAGE_METADATA_TYPE, refName: apexPage.apexPage }),
+    instancesIndex.byTypeAndApiName,
+  )
 
-  const recordTypeRefs = awu(addedOrModifiedInstancesChangesByType[RECORD_TYPE_METADATA_TYPE] ?? [])
-    .flatMap(async recordType => refsFromProfileOrPermissionSet(
-      profilesAndPermSetsChanges,
-      recordType,
-      RECORD_TYPE_SECTION,
-    ))
+  const objectRefs = await instanceRefsFromProfileOrPermissionSet(
+    profilesAndPermSetsChanges,
+    OBJECT_SECTION,
+    (object, key) => ({ typeName: key, refName: object.object }),
+    instancesIndex.byTypeAndApiName,
+  )
 
-  const objectRefs = awu(customObjectChanges)
-    .flatMap(async object => refsFromProfileOrPermissionSet(profilesAndPermSetsChanges, object, OBJECT_SECTION))
-
-  const recordTypesByApiName = await awu(addedOrModifiedInstancesChangesByType[RECORD_TYPE_METADATA_TYPE] ?? [])
-    .keyBy(async recordType => (await safeApiName(getChangeData(recordType))) ?? '')
+  const fieldPermissionsRefs = awu(relevantFieldChanges)
+    .flatMap(async field => fieldRefsFromProfileOrPermissionSet(profilesAndPermSetsChanges, field))
 
-  const recordTypeRefsFromLayouts = awu(addedOrModifiedInstancesChangesByType[LAYOUT_TYPE_ID_METADATA_TYPE] ?? [])
-    .map(getChangeData)
-    .flatMap(async layout => {
-      const apiName = await safeApiName(layout)
-      if (apiName === undefined) {
-        return []
-      }
-      return profilesAndPermSetsChanges
-        .filter(profileOrPermissionSet => _.get(getChangeData(profileOrPermissionSet).value[LAYOUTS_SECTION], apiName))
-        .flatMap(profileOrPermissionSet => (
-          Object.values(getChangeData(profileOrPermissionSet).value[LAYOUTS_SECTION])
-            .flatMap(layoutAssignments => recordTypeRefsFromLayoutAssignments(
-              layoutAssignments,
-              recordTypesByApiName,
-              profileOrPermissionSet,
-              apiName
-            ))
+  const recordTypeRefs = profilesAndPermSetsChanges
```

Comment:
Why can't we Dry this with a simple instanceRefsFromProfileOrPermissionSet like the other example. Doesn't the instanceNameGetter need to be slightly more complicated and everything else be the same?

Reply:
You'd have to change `instanceNameGetter` to return an array, and if I'm going to do that, I may as well follow your suggestion above and extract the `recordType` refs at the same time I extract the layout refs.

Reply:
We made it return an array, wouldn't it be simpler to use profilesAndPermSetsChanges with a "specific" RefNameGetter?

Reply:
I don't think so, because recordTypeVisibilities is nested 2-deep, so the reference source behaves differently from all other cases:

```
  recordTypeVisibilities = {
    Case = {
      SomeCaseRecordType = {
        default = true
        recordType = salesforce.RecordType.instance.Case_SomeCaseRecordType
        visible = true
      }
      SomeOtherCaseRecordType = {
        default = false
        recordType = salesforce.RecordType.instance.Case_SomeOtherCaseRecordType
        visible = true
      }
    }
...
}
```

Reply:
Doesn't something along the lines of the following work?
```suggestion
  (recordTypeVisibiliy, key) => Object.entries(recordTypeVisibilities).map(([someType, { recordType }]) => ({ typeName: key, refName: recordType })
```